diff --git a/conf/nginx.conf b/conf/nginx.conf
index 29bc085..8e4fbd5 100644
--- a/conf/nginx.conf
+++ b/conf/nginx.conf
@@ -1,3 +1,5 @@
+master_process off;
+daemon off;
 
 #user  nobody;
 worker_processes  1;
@@ -11,6 +13,8 @@ worker_processes  1;
 
 events {
     worker_connections  1024;
+    use select;
+    multi_accept off;
 }
 
 
diff --git a/src/os/unix/ngx_process_cycle.c b/src/os/unix/ngx_process_cycle.c
index 5bc5ce9..e1ad353 100644
--- a/src/os/unix/ngx_process_cycle.c
+++ b/src/os/unix/ngx_process_cycle.c
@@ -294,6 +294,8 @@ ngx_single_process_cycle(ngx_cycle_t *cycle)
         }
     }
 
+    static volatile int run_count = 0;
+
     for ( ;; ) {
         ngx_log_debug0(NGX_LOG_DEBUG_EVENT, cycle->log, 0, "worker cycle");
 
@@ -328,6 +330,9 @@ ngx_single_process_cycle(ngx_cycle_t *cycle)
             ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "reopening logs");
             ngx_reopen_files(cycle, (ngx_uid_t) -1);
         }
+
+        if (run_count >= 1) exit(0);
+        run_count += 1;
     }
 }
 
